test/extensions/tracers/opencensus/tracer_test.cc

```diff
@@ -0,0 +1,134 @@
+#include <iostream>
+#include <vector>
+
+#include "envoy/config/trace/v2/trace.pb.validate.h"
+
+#include "extensions/tracers/opencensus/opencensus_tracer_impl.h"
+
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/tracing/mocks.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "opencensus/trace/exporter/span_data.h"
+#include "opencensus/trace/exporter/span_exporter.h"
+#include "opencensus/trace/span_id.h"
+
+using testing::NiceMock;
+using testing::Return;
+
+namespace opencensus {
+namespace trace {
+namespace exporter {
+
+class SpanExporterTestPeer {
+public:
+  static constexpr auto& exportForTesting = SpanExporter::ExportForTesting;
+};
+
+} // namespace exporter
+} // namespace trace
+} // namespace opencensus
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace OpenCensus {
+
+using ::opencensus::trace::exporter::SpanData;
+using ::opencensus::trace::exporter::SpanExporter;
+
+namespace {
+
+// Custom export handler. We register this as an OpenCensus trace exporter, and
+// use it to catch the spans we produce.
+class SpanCatcher : public SpanExporter::Handler {
+public:
+  void Export(const std::vector<SpanData>& spans) override {
+    absl::MutexLock lock(&mu_);
+    for (const auto& span : spans) {
+      spans_.emplace_back(span);
+    }
+  }
+
+  // Returns generated SpanData, and clears the catcher.
+  std::vector<SpanData> catchSpans() {
+    // OpenCensus's trace exporter is running in a background thread, waiting
+    // for a periodic export. Force it to flush right now.
+    opencensus::trace::exporter::SpanExporterTestPeer::exportForTesting();
+    absl::MutexLock lock(&mu_);
+    std::vector<SpanData> ret = std::move(spans_);
+    spans_.clear();
+    return ret;
+  }
+
+private:
+  mutable absl::Mutex mu_;
+  std::vector<SpanData> spans_ GUARDED_BY(mu_);
+};
+
+// Use a Singleton SpanCatcher.
+SpanCatcher* getSpanCatcher() {
+  static auto g_span_catcher = new SpanCatcher();
+  return g_span_catcher;
+}
+
+// Call this before generating spans to register the exporter that catches them.
+void registerSpanCatcher() {
+  static bool done_once = false;
+  if (done_once) {
+    return;
+  }
+  SpanExporter::RegisterHandler(absl::WrapUnique(getSpanCatcher()));
+  done_once = true;
+}
+
+} // namespace
+
+// Create a Span via the driver, test all of the Tracing::Span API, and verify
+// the produced SpanData.
+TEST(OpenCensusTracerTest, Span) {
+  registerSpanCatcher();
+  envoy::config::trace::v2::OpenCensusConfig oc_config;
+  std::unique_ptr<Tracing::Driver> driver_(new OpenCensus::Driver(oc_config));
+
+  NiceMock<Tracing::MockConfig> config_;
+  Http::TestHeaderMapImpl request_headers_{
+      {":path", "/"}, {":method", "GET"}, {"x-request-id", "foo"}};
+  const std::string operation_name_{"my_operation_1"};
```

Comment:
Nit: `operation_name` (this isn't a member), same goes with above field names.

Reply:
Done.